MEN-5718: Fix broken document migration

alfrunes opened the pull request:
Fixes the document migration for collection with more than 255 documents and applies the migration to legacy (multiple namespaces) layout.

merlin-northern commented on store/mongo/migration_2_0_0.go:
```diff
@@ -148,6 +139,7 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
 				writes = append(writes, mongo.NewInsertOneModel().SetDocument(item))
 			}
 			if len(writes) == findBatchSize {
+				_, err = collOut.BulkWrite(ctx, writes)
```
oh wow, what a shame. I actually for got to call Write, the error handling is there, I simply did not put the Write here. I am too ashamed and not qualified to review. looks good.

alfrunes replied:
I'm more surprised that it passed the testing infrastructure. I think there's some room for improvement when testing database migrations.
In any case, I'd appreciate a review from you, if you have the time.

alfrunes commented:
@mender-test-bot start pipeline

merlin-northern commented on store/mongo/migration_2_0_0.go:
```diff
@@ -104,50 +93,59 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
-		if err != nil {
-			return err
+		if len(idxes.Indexes) > 0 {
+			_, err := collOut.Indexes().CreateMany(ctx, collections[collection].Indexes)
```
if m.db == DbName then coll is collOut, so do we need to create them in any case? shall we remove the indices and then create them?

alfrunes replied:
Oops.. This was only supposed to execute when `m.db == DbName`. We don't have to drop existing indexes if the keys and index names don't conflict with the ones created here.
Good catch, I will move this block inside the next if clause.

alfrunes commented on store/mongo/migration_2_0_0.go:
```diff
@@ -43,18 +43,7 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
-				},
-			},
-		},
+		DevicesCollectionName: {},
```
There is a unique index on the `_id` field by default. I couldn't find any queries that retrieves all the documents for a given tenant so I'm removing this index to save the memory. We could recreate it later if needed

alfrunes commented on store/mongo/migration_2_0_0.go:
```diff
@@ -104,50 +93,59 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
+			itemD := mstore.WithTenantID(ctx, item)
+			writes = append(writes, mongo.
+				NewReplaceOneModel().
+				SetFilter(bson.D{{Key: "_id", Value: item["_id"]}}).
```
I was just lazy :sloth:

alfrunes replied:
Fixed ;)

alfrunes commented on store/mongo/migration_2_0_0.go:
```diff
@@ -104,50 +93,59 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
 		}
 		defer cur.Close(ctx)
 
-		writes := make([]mongo.WriteModel, 0, findBatchSize)
-
 		// migrate the documents
 		for cur.Next(ctx) {
-			item := bson.D{}
+			var item bson.M
```
I changed it to a map so I wouldn't have to recreate it later when I retrieved the `_id` key below. In general I would prefer `bson.D` because of the lower complexity of the data type.

alfrunes replied:
Reverted to `bson.D` and used the raw bson document to lookup the `_id` field.

alfrunes commented on store/mongo/migration_2_0_0.go:
```diff
@@ -104,50 +93,59 @@ func (m *migration_2_0_0) Up(from migrate.Version) error {
+				SetFilter(bson.D{{Key: "_id", Value: item["_id"]}}).
+				SetUpsert(true).
+				SetReplacement(itemD))
 			if len(writes) == findBatchSize {
+				_, err = collOut.BulkWrite(ctx, writes)
 				if err != nil {
 					return err
 				}
```
In the current form, the test is using `findBatchSize+1` documents in both the primary and a tenant database, which covers both paths with batch replacements and remainders. Could you elaborate why we need the above document counts?

merlin-northern replied:
it explores  all the cases, looking at it as in black box testing. having them would make me sleep in peace :) if you feel that +1 scenario is enough, I will not block it.

alfrunes commented:
Thanks for the review @merlin-northern. While addressing your comments, I also realized that it doesn't make sense to run the migration twice if `2.0.0` was not applied, so I simply bumped the version and removed the former `2.0.0` version of the migration.

alfrunes commented:
@mender-test-bot start pipeline

merlin-northern commented on store/mongo/migrations.go:
```diff
@@ -46,26 +47,37 @@ func Migrate(ctx context.Context,
+				client: client,
+				db:     dbName,
+			},
+			// NOTE: Future migrations need only be applied to DbName
```
but not the `dbName` ;-) I would somehow underline that. or add another comment after the loop.

merlin-northern commented on store/mongo/migrations.go:
```diff
@@ -20,6 +20,7 @@ import (
+	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
 )
 
```
`DbVersion = "2.0.0"` is this correct?

alfrunes replied:
That is indeed an error. I bumped it to 2.0.1.

merlin-northern reviewed (approved):
lgtm. I hate the formatting there: https://github.com/mendersoftware/deviceconnect/pull/229/files#diff-2cce0784f2652ff85b2d3fa8f93a7d05871f927adbc514d96d8263ad7c3bf756R117 would add new lines after every curly bracket. but up to you.
